### ISSUE TITLE
test.py is reporting around 7% higher AP50 compared to COCOeval

### ISSUE THREAD

**Issue by deepakcrk** [ultralytics/yolov5#1302]
## ❔Question
While calculating AP50 using test.py, it's showing around 7% higher AP50 than COCOeval function in the pycotools.
ap_per_class function is reporting around 7% higher AP50

**Comment by glenn-jocher** [comment]
@deepakcrk thanks for the FYI. Can you provide screenshots of what you see?

**Comment by deepakcrk** [comment]
@glenn-jocher 

I have attached results on a small dataset (150 images), The results are from finetuned yolov5 model (only 1 class).

The command used is: _python3 test.py --weights person_single_class.pt --task test --single-cls_
Screenshot of yolov5s result:

![screeen3](https://user-images.githubusercontent.com/10599036/98320341-bac17580-2008-11eb-9077-2f25e9ba3c19.png)


**Result obtained from _Object-Detection-Metrics_ repo:**
![object_detetion](https://user-images.githubusercontent.com/10599036/98320437-f65c3f80-2008-11eb-8391-c3d1b581ddb0.PNG)


**Result obtained from COCOeval API:**

 _Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.629_

**Comment by glenn-jocher** [comment]
@deepakcrk thanks. The difference may be due to a recent change in mAP computation in PE https://github.com/ultralytics/yolov5/pull/1206. Can you try to replace these lines:
https://github.com/ultralytics/yolov5/blob/fed5d9cdfb25ffe046062f0fb220cb20fc78ca42/utils/general.py#L331-L334

with this code and see if this creates a better match?
```python
    # Append sentinel values to beginning and end
    mrec = np.concatenate(([0.], recall, [recall[-1] + 1E-3]))
    mpre = np.concatenate(([0.], precision, [0.]))

```

**Comment by deepakcrk** [comment]
@glenn-jocher with above change AP@50 reduced to 68.0%  (from 68.1%)

Still, it's showing around 5.1% deviation from COCOeval AP.

**Comment by glenn-jocher** [comment]
@deepakcrk hmm ok thanks. --single-cls may play a part, it's only intended for multi-class datasets that you want to train and test as a single class. If your dataset already only contains a single class then it's redundant. You might want to test without it.

**Comment by glenn-jocher** [comment]
@deepakcrk also keep in mind, the repo mAP follows the COCO computation method, with 101 interpolation points over the P-R curves. Pascal uses a different method for handling the PR curves I believe.
https://github.com/ultralytics/yolov5/blob/fed5d9cdfb25ffe046062f0fb220cb20fc78ca42/utils/general.py#L338-L346

**Comment by deepakcrk** [comment]
@glenn-jocher 

I have trained as a single class problem, by modifying nc=1 in the yaml file. Also tested with and without _--single-cls_ , and was getting the same result with both (as I have only one class)


I have tested AP50, As per pascal VOC, and by using pycocotool, both were giving almost the same AP (63.27 and 62.9) but by using test.py I was getting 68.0%

**Comment by glenn-jocher** [comment]
@deepakcrk ok got it. Are you sure you are using the same predictions for each mAP run? 

Are you providing an annotations.json file for test.py to use with pycocotools or how are you using pycocotools?

**Comment by deepakcrk** [comment]
@glenn-jocher 

The problem is not with AP calculation..

In the _test.py_, even though the default --img-size is **640**, its passing an image size of **672** to the model.. (looks like its not using default size mentioned in the argparser..)

And for pycocotool(COCOeval), I was passing results from _detect.py_, where there is no mismatch between default size and the size internally used.. (So basically I was using different input sizes for AP evaluation earlier)

And there may be a slight difference in the input image creation between detect.py and test.py which may be caused further deviation from AP.

**Comment by glenn-jocher** [comment]
@deepakcrk ah, ok got it! Yes, test.py has a pad variable that it sends to the dataloader. Default pad=0.5 produces half a stride (16 pixels) padding on each side. You can remove the padding here:
https://github.com/ultralytics/yolov5/blob/ea7e78cb1159e6a17821772c85c4c23ccc823b16/test.py#L94-L96
